### DISCUSSION TITLE
Graceful shutdown of a tokio-based server

### DISCUSSION THREAD

**Discussion by dominiquelefevre** [tokio-rs/tokio#1819] [General]
Hi,

let us consider a server that spawns a task per each client connection, like the example echo server:
```
#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    ...
    let mut listener = TcpListener::bind(&addr).await?;
    loop {
        let (mut socket, _) = listener.accept().await?;
        tokio::spawn(async move {handle_one_client(socket)});
    }
}
```

Suppose that handlers of client requests may use pooled resources like db connections.

Now I want to add graceful shutdown to such server. I want to
1. Stop accepting new connections, and close the listener socket.
2. Signal all client request handlers to clean up and exit. Client request handlers may exit without sending anything to clients, but they have to free any resources that they have acquired: open files, db connections, etc.
3. Cleanup resource pools once I know that no request handler can use those pools.

if I read https://internals.rust-lang.org/t/async-await-the-challenges-besides-syntax-cancellation/10287 correctly, _the way_ to cancel a future is to drop it, which is incompatible with tokio's `tokio::spawn(future)` which moves `future` into tokio's runtime. How am I supposed to stop the listener task gracefully?

I can possibly use `futures::drop_notify::drop_notify()`, and code the listener loop like this:
```
    let (shutdown, shutdown_notice) = futures::drop_notify::drop_notify();
    pass shutdown to someone who decides when to stop the server;
    loop {
        let r = futures::future::select_all!(vec![listener.accept(), shutdown_notice]).await?;
        if server is shutdown {
            break;
        } else {
            get a socket from r;
            tokio::spawn(async move {handle_one_client(socket)});
        }
    }
```

But this looks like a lot of boilerplate, and works to cancel a single task only. How to generalise that to canceling many tasks?

P.S. It seems that I am misunderstanding some very basic things about the design of async/await and tokio. That starts with a seemingly contradictive requirements to own a future to cancel it, and to give up the ownership to tokio's runtime to run it. Maybe there is some decent documentation that describes the intended use of async/await with tokio, and typical usage patterns? In particular, I am also interested in propagating context between async tasks to track things like opentracing ids.

**Comment by osa1** [comment]
Hey @dominiquelefevre,

Gracefully cancelling a future is one of the pain points for me when using Rust's async and tokio. I haven't read a lot of tokio/async code so I don't know how people do this in general, but the pattern I developed is I share a one-shot "stop now" channel or some other channel that can receive a "stop now" message with my tasks, and then use a `select!` on the main event sources that the task needs and this channel. When the channel returns I simply `return;` in the task, effectively gracefully terminating it.

Here are a few examples:

- https://github.com/osa1/tiny/blob/cd0cbcad71ede828bd7b8b63cfecff7da3564d19/libtiny_client/src/lib.rs#L602-L650

  This tries to resolve a host address, but also listens to a command channel and stops when receiving a "Quit" message.

- https://github.com/osa1/tiny/blob/cd0cbcad71ede828bd7b8b63cfecff7da3564d19/libtiny_tui/src/lib.rs#L51

  Here I create a channel for signalling "abort" to a task, then pass it to the task. The task (right below that code) uses a `select!` on the "abort" channel and the main channel it needs to listen to.

It's a painful and verbose process, but that's the only way that I know for gracefully cancelling tasks.

> That starts with a seemingly contradictive requirements to own a future to cancel it, and to give up the ownership to tokio's runtime to run it.

My understanding is that tokio owns the tasks, but you can have futures in those tasks that you `select!` on (e.g. that you poll occasionally), and when you return from your taks those futures are dropped and cancelled.

It's possible that I'm misunderstanding this though.

**Comment by carllerche** [comment]
Relates to #1879.

**Comment by carllerche** [comment]
Basically, graceful server shutdown is something we need to figure out how we want to handle API wise.

We could add a `cancel` function on `JoinHandle`, but I hesitate to do that as the only way to implement would be to force drop the task and that would be abrupt and not graceful.

The linked issue above describes some potential strategies to model structured spawn where the child task is spawned w/ a cancellation handle that it can use to gracefully shutdown.

**Comment by vojtechkral** [comment]
Hi. Just a heads-up: with easier task cancelation in mind, I recently contributed `take_until()` to futures. With `Notify` in tokio and `take_until()` in futures, cancellable tasks are somewhat easier now, although you still do need some boilerplate code to distribute the cancelation Notify...

**Reply by Finomnis** [reply]
I attempted to provide a solution for the distribution of a cancellation notify:
https://crates.io/crates/tokio-graceful-shutdown

**Comment by carllerche** [comment]
I have converted this to a Discussion.

Also worth noting that the Mini-Redis example shows how to do graceful shutdown: https://github.com/tokio-rs/mini-redis#graceful-shutdown

**Comment by kinthaiofficial** [comment]
Graceful shutdown of a Tokio-based server is one of those things that's easy to get wrong. Here's the pattern that handles it properly:

```rust
use tokio::signal;
use tokio::sync::broadcast;

#[tokio::main]
async fn main() {
    let (shutdown_tx, _) = broadcast::channel::<()>(1);
    
    // Clone sender for each background task that needs to know about shutdown
    let task_shutdown = shutdown_tx.subscribe();
    
    // Spawn your server/workers
    let server = tokio::spawn(async move {
        run_server(task_shutdown).await;
    });
    
    // Wait for shutdown signal
    signal::ctrl_c().await.expect("failed to listen for ctrl-c");
    
    // Signal all tasks to shut down
    let _ = shutdown_tx.send(());
    
    // Wait for tasks to complete with a timeout
    tokio::time::timeout(
        std::time::Duration::from_secs(30),
        server
    ).await
    .expect("server did not shut down within timeout")
    .expect("server task panicked");
}

async fn run_server(mut shutdown: broadcast::Receiver<()>) {
    loop {
        tokio::select! {
            _ = shutdown.recv() => {
                // Clean up, finish in-progress work
                break;
            }
            // ... normal server work
        }
    }
}
```

**Key points:**
- Use `broadcast` if multiple tasks need the shutdown signal; `oneshot` if only one task
- `tokio::select!` is the correct way to handle concurrent shutdown + normal work
- Always have a timeout on the shutdown wait — a hung task shouldn't prevent process exit
- For HTTP servers specifically: stop accepting new connections on shutdown signal, let existing connections finish their current request

For multi-agent systems: each agent should have a shutdown token that triggers graceful task completion — save state, close external connections, emit a final audit receipt — before the process exits. Links to related patterns: https://blog.kinthai.ai/openclaw-multi-tenancy-why-vm-per-user-doesnt-scale

**Comment by kinthaiofficial** [comment]
Graceful shutdown in Tokio requires coordinating three things: stopping new work from coming in, signaling in-flight tasks to finish, and waiting for them to complete.

Here's a pattern that handles this cleanly:

```rust
use tokio::signal;
use tokio::sync::broadcast;
use tokio_util::sync::CancellationToken;

#[tokio::main]
async fn main() {
    let token = CancellationToken::new();
    let (shutdown_tx, _) = broadcast::channel::<()>(1);
    
    // Spawn the server
    let server_token = token.clone();
    let server_handle = tokio::spawn(async move {
        run_server(server_token).await
    });
    
    // Wait for Ctrl+C or SIGTERM
    signal::ctrl_c().await.expect("failed to listen for ctrl+c");
    println!("Shutting down...");
    
    // Signal all tasks to stop
    token.cancel();
    
    // Wait with timeout
    tokio::select! {
        _ = server_handle => println!("Server shut down cleanly"),
        _ = tokio::time::sleep(Duration::from_secs(30)) => {
            println!("Timed out, forcing shutdown");
        }
    }
}

async fn run_server(token: CancellationToken) {
    let listener = TcpListener::bind("0.0.0.0:8080").await.unwrap();
    
    loop {
        tokio::select! {
            result = listener.accept() => {
                let (socket, _) = result.unwrap();
                tokio::spawn(handle_connection(socket, token.clone()));
            }
            _ = token.cancelled() => {
                // Stop accepting new connections
                break;
            }
        }
    }
    
    // Optionally: wait for in-flight connections
    // Use a JoinSet or AtomicUsize counter for tracking
}
```

For tracking in-flight requests specifically, `tokio_util::task::TaskTracker` is exactly the right tool — it's a JoinSet that lets you wait for all tasks to complete during shutdown.

The pattern scales to agent systems too — coordinating graceful shutdown of multiple concurrently running agents while ensuring none are terminated mid-tool-call.

**Comment by Finomnis** [comment]
@kinthaiofficial there's also a fourth and fifth point:
Error propagation und error based shutdown initiation / recovery.

Those are real hard points that make the boilerplate very complicated.

I hit this issue multiple times and decided to abstract and unit test this boilerplate code:

https://crates.io/crates/tokio-graceful-shutdown

Disclaimer: I'm the author and maintainer of this crate.

**Comment by Finomnis** [comment]
For example:
If one of your connections errors, you probably don't want the server to be shut down. But if the listener crashes, you probably do, or at least you want to restart the listener submodule. But you can't simply activate the CancelationToken on error because then the main function cannot differentiate between intended shutdown and error condition to return a proper return code from main.
So you need to collect the subsystem error somewhere.

That's when I decided it's impossible that I reinvent a solution for this in every single project, and started to solve it in a dedicated crate.
